### PR TITLE
Better error message for quantized dispatch

### LINKED ISSUE #28518: [quantization] Dispatch key error message is confusing to users
For example: https://github.com/pytorch/pytorch/issues/28483

User encounters the following inscrutable error message:

```
RuntimeError: Didn't find kernel to dispatch to for operator 'quantized::conv2d'. Tried to look up kernel for dispatch key 'CPUTensorId'. Registered dispatch 
keys are: [QuantizedCPUTensorId] (lookup_ at /opt/conda/conda-bld/pytorch_1570910687650/work/aten/src/ATen/core/dispatch/DispatchTable.h:249)
```

We should figure out a way to print a more friendly error message, such as:

```
Tried to run quantized::conv2d, but the input was not quantized! Please ensure you have QuantStub during model conversion, or you explicitly call  one of quantize_per_tensor, quantize_per_channel
```

cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28635 Better error message for quantized dispatch**

Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

### CODE REVIEW

File: aten/src/ATen/core/dispatch/DispatchTable.h [hunk 1]
```diff
@@ -243,7 +243,19 @@ class DispatchTable final {
               "Available functions are ", listAllDispatchKeys())
       }
 
-      const std::string dispatch_key_str = dispatch_key.has_value() ? toString(*dispatch_key) : "None";
+      if (dispatch_key.value() == TensorTypeId::QuantizedCPUTensorId) {
+        // If the input is quantized, but the quantization is not supported.
+        TORCH_CHECK(false, "Tried runninig '", operator_name_, "', but it",
```
Comment (jamesr66a) [inline aten/src/ATen/core/dispatch/DispatchTable.h]: s/runninig/running/

Comment (jamesr66a) [reply aten/src/ATen/core/dispatch/DispatchTable.h]: You might also want to specify something like "Tried running X *with a quantized tensor* but X expects a non-quantized input" just for clarity

File: aten/src/ATen/core/dispatch/DispatchTable.h [hunk 2]
```diff
@@ -243,7 +243,19 @@ class DispatchTable final {
...
+                    " expects non-quantized input. Please, check the",
+                    " documentation for supported kernels.");
+      }
+      // If the input is not quantized, but the kernel is.
+      TORCH_CHECK(!kernels_.lookup(TensorTypeId::QuantizedCPUTensorId), "Tried",
```
Comment (jamesr66a) [inline aten/src/ATen/core/dispatch/DispatchTable.h]: For consistency, can we pull this conditional out into an `if` statement to make it clear this is not the fall-through path? I was confused when first reading this

**Review by jamesr66a — COMMENTED** [review]
Thanks! Left some small comments

cc @smessmer

File: aten/src/ATen/core/dispatch/DispatchTable.h [hunk 3]
```diff
@@ -243,7 +243,19 @@ class DispatchTable final {
               "Available functions are ", listAllDispatchKeys())
       }
 
-      const std::string dispatch_key_str = dispatch_key.has_value() ? toString(*dispatch_key) : "None";
+      if (dispatch_key.value() == TensorTypeId::QuantizedCPUTensorId) {
```
Comment (jerryzh168) [inline aten/src/ATen/core/dispatch/DispatchTable.h]: do you need to add `has_value()` check?

Comment (jamesr66a) [reply aten/src/ATen/core/dispatch/DispatchTable.h]: That's already checked on line 238

**Review by jianyuh — APPROVED** [review]
LGTM!

**Comment by smessmer** [comment]
I don't think we want this. The core dispatch logic is agnostic of quantization and should stay this way. Quantization depends on dispatch, not the other way round.

**Pull request closed without merging**